# Отчет

## Импорт библиотек и открытие данных

In [1]:
import numpy as np
from prepare_report import knapsack_report, salesman_report
from prepare_report import res_table_salesman, res_table_knapsack
from data.get_branches import get_salesman, get_knapsack

In [ ]:
!unzip /content/genetic

In [2]:
capacity, profits, weights, optimal = get_knapsack('data/benchmarks/')
graph = get_salesman('data/output_graphs.npy')

## Результаты knapsack 0-1

In [ ]:
total_time = []
for i in range(10):
    time_knapsack, weight_knapsack,profit_knapsack, id_knapsack = knapsack_report(capacity, weights, profits)
    total_time.append(time_knapsack)

In [ ]:
total_time = np.array(total_time)
total_time = np.mean(total_time, axis=0)

In [ ]:
res_table_knapsack(sorted(capacity.keys()), total_time, 
                   weight_knapsack, profit_knapsack, 
                   id_knapsack)

p01              p02                 p03  \
Weight                             165               26                 190   
Profit                             309               51                 150   
Time                          0.079288         0.061499            0.061475   
ID      [1, 1, 1, 1, 0, 1, 0, 0, 0, 0]  [0, 1, 1, 1, 0]  [1, 1, 0, 0, 1, 0]   

                          p04                       p05  \
Weight                     50                       104   
Profit                    107                       900   
Time                 0.067658                  0.064515   
ID      [1, 0, 0, 1, 0, 0, 0]  [1, 0, 1, 1, 1, 0, 1, 1]   

                          p06                                            p07  
Weight                    169                                            749  
Profit                   1735                                           1458  
Time                 0.067684                                       0.080929  
ID      [0, 1, 0, 1, 0, 0, 1]  [1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1]

**Предыдущая таблица (лаб2)**

| Рюкзаки\Ответ 	| Вес 	| Стоимость 	| Время 	| ID 	|
|---|---|---|---|---|
| p01 	| 165   	| 309   	| 0.001335 	| [1, 1, 1, 1, 0, 1, 0, 0, 0, 0] 	|
| p02 	| 26    	| 51    	| 0.000094 	| [0, 1, 1, 1, 0] 	|
| p03 	| 190   	| 150    	| 0.001033  	| [1, 1, 0, 0, 1, 0] 	|
| p04 	| 50    	| 107    	| 0.000238  	| [1, 0, 0, 1, 0, 0] 	|
| p05 	| 104   	| 900   	| 0.000658  	| [1, 0, 1, 1, 1, 0, 1, 1] 	|
| p06 	| 169   	| 1735   	| 0.000863 	| [0, 1, 0, 1, 0, 0, 1] 	|
| p07 	| 749   	| 1458   	| 0.010723 	| [1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1] 	|

## Результаты коммивояжер

In [7]:
sorted_keys = ['gr17', 'bays29', 'att48', 'ch150', 'a280', 'fl417']
# paramet = {'gr17': (100, 3_000_000, 10, 0.2), 'bays29': (100, 3_000_000, 10, 0.2), 
#           'att48': (100, 2_000_000, 10, 0.2), 'ch150': (100, 1_000_000, 10, 0.2),
#           'a280': (100, 700_000, 10, 0.2), 'fl417': (100, 500_000, 10, 0.2)}
paramet = (100, 10000, 20)

In [8]:
time_salesman = []
weight_salesman = []
way_salesman = []
answer_salesman = []

In [9]:
for i in sorted_keys:
    result = salesman_report(graph[i], paramet) 
    
    time_salesman.append(result[3])
    weight_salesman.append(result[1])
    way_salesman.append(result[0])
    answer_salesman.append(result[2])
    
    print(f"{i}: {result[1]} vs {result[2]}")

100%|████████████████████████████████████| 10000/10000 [00:21<00:00, 466.86it/s]


gr17: 2085.0 vs 2085.0


100%|████████████████████████████████████| 10000/10000 [00:29<00:00, 343.14it/s]


bays29: 2187.0 vs 2020.0


100%|████████████████████████████████████| 10000/10000 [00:41<00:00, 242.52it/s]


att48: 12358.0 vs 10628.0


100%|█████████████████████████████████████| 10000/10000 [01:42<00:00, 97.51it/s]


ch150: 10686.84236073494 vs 6528.0


100%|█████████████████████████████████████| 10000/10000 [02:54<00:00, 57.35it/s]


a280: 2799.3609828948975 vs 2579.0


100%|█████████████████████████████████████| 10000/10000 [03:50<00:00, 43.35it/s]


fl417: 29613.841302871704 vs 11861.0


In [10]:
res_table_salesman(sorted_keys, time_salesman, 
          weight_salesman, answer_salesman, way_salesman)

gr17   
Weight                                                  2085.0  \
True answer                                             2085.0   
Time                                                 21.717152   
Way          [7, 6, 12, 3, 0, 15, 11, 8, 4, 1, 9, 10, 2, 14...   

                                                        bays29   
Weight                                                  2187.0  \
True answer                                             2020.0   
Time                                                 30.144583   
Way          [0, 4, 2, 28, 25, 8, 11, 5, 27, 7, 22, 6, 24, ...   

                                                         att48   
Weight                                                 12358.0  \
True answer                                            10628.0   
Time                                                 44.700244   
Way          [18, 36, 5, 27, 6, 17, 43, 30, 37, 7, 0, 8, 39...   

                                                         ch150   
Weight                                            10686.842361  \
True answer                                             6528.0   
Time                                                147.468251   
Way          [76, 143, 128, 30, 26, 116, 139, 59, 65, 16, 1...   

                                                          a280   
Weight                                             2799.360983  \
True answer                                             2579.0   
Time                                                218.401376   
Way          [1, 279, 0, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 1...   

                                                         fl417  
Weight                                            29613.841303  
True answer                                            11861.0  
Time                                                267.469919  
Way          [7, 0, 9, 15, 20, 19, 12, 1, 3, 2, 8, 10, 14, ...

In [7]:
res_table_salesman(sorted_keys, time_salesman, 
          weight_salesman, answer_salesman, way_salesman)

gr17   
Weight                                                  2348.0  \
True answer                                             2085.0   
Time                                               4466.672489   
Way          [8, 11, 15, 3, 6, 0, 5, 13, 2, 10, 4, 1, 9, 14...   

                                                        bays29   
Weight                                                  3620.0  \
True answer                                             2020.0   
Time                                               6132.544323   
Way          [24, 6, 22, 23, 18, 26, 7, 5, 11, 9, 15, 27, 0...   

                                                         att48   
Weight                                                 30250.0  \
True answer                                            10628.0   
Time                                               5998.871162   
Way          [39, 37, 7, 4, 9, 38, 44, 34, 25, 24, 13, 33, ...   

                                                         ch150   
Weight                                            43557.043917  \
True answer                                             6528.0   
Time                                                7775.04206   
Way          [64, 76, 88, 125, 27, 108, 41, 16, 23, 92, 32,...   

                                                          a280   
Weight                                             2818.621637  \
True answer                                             2579.0   
Time                                              10368.859275   
Way          [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...   

                                                         fl417  
Weight                                            55387.192585  
True answer                                            11861.0  
Time                                              10359.553895  
Way          [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...

In [8]:
for i in way_salesman:
    print(i)

[8, 11, 15, 3, 6, 0, 5, 13, 2, 10, 4, 1, 9, 14, 16, 7, 12]
[24, 6, 22, 23, 18, 26, 7, 5, 11, 9, 15, 27, 0, 13, 19, 17, 21, 10, 1, 2, 28, 25, 20, 4, 8, 12, 3, 16, 14]
[39, 37, 7, 4, 9, 38, 44, 34, 25, 24, 13, 33, 28, 2, 27, 31, 23, 47, 22, 0, 21, 8, 14, 43, 42, 36, 26, 45, 18, 5, 30, 17, 15, 3, 1, 41, 20, 6, 10, 35, 16, 11, 12, 32, 40, 46, 19, 29]
[64, 76, 88, 125, 27, 108, 41, 16, 23, 92, 32, 29, 50, 114, 137, 110, 17, 140, 28, 57, 73, 129, 102, 53, 7, 78, 107, 67, 148, 77, 5, 101, 98, 113, 39, 144, 54, 139, 122, 132, 79, 45, 115, 26, 66, 49, 4, 130, 142, 149, 95, 133, 63, 6, 75, 14, 136, 46, 83, 141, 44, 146, 60, 42, 123, 138, 3, 43, 147, 2, 22, 15, 48, 121, 30, 145, 21, 103, 127, 91, 33, 105, 10, 85, 56, 11, 80, 126, 37, 89, 35, 68, 47, 58, 1, 51, 25, 69, 18, 36, 117, 38, 116, 134, 59, 31, 100, 71, 24, 84, 19, 111, 90, 97, 62, 120, 20, 65, 128, 135, 8, 119, 131, 12, 87, 86, 0, 106, 96, 72, 99, 34, 94, 61, 124, 13, 104, 40, 143, 70, 118, 93, 112, 55, 74, 109, 82, 81, 9, 52]
[0, 1, 2, 